clients/git/client.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2021 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package git defines helper functions for clients.RepoClient interface.
+package git
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	cp "github.com/otiai10/copy"
+
+	"github.com/ossf/scorecard/v4/clients"
+)
+
+const repoDir = "repo*"
+
+var (
+	errNilCommitFound = errors.New("nil commit found")
+	errEmptyQuery     = errors.New("query is empty")
+)
+
+type Client struct {
+	gitRepo        *git.Repository
+	worktree       *git.Worktree
+	listCommits    *sync.Once
+	tempDir        string
+	errListCommits error
+	commits        []clients.Commit
+	commitDepth    int
+}
+
+func (c *Client) InitRepo(uri, commitSHA string, commitDepth int) error {
+	// cleanup previous state, if any.
+	c.Close()
+	c.listCommits = new(sync.Once)
+	c.commits = nil
+
+	// init
+	c.commitDepth = commitDepth
+	tempDir, err := os.MkdirTemp("", repoDir)
+	if err != nil {
+		return fmt.Errorf("os.MkdirTemp: %w", err)
+	}
+
+	// git clone
+	const filePrefix = "file://"
+	if strings.HasPrefix(uri, filePrefix) {
+		if err := cp.Copy(strings.TrimPrefix(uri, filePrefix), tempDir); err != nil {
```

Comment:
if this is being called on a temp directory already (from another scorecard check perhaps), does it make sense to have a way to avoid this extra copy?

Reply:
The expectation here is that this will be called when `--local` CLI flag is used by the client (never from another Scorecard check). In that case, its easiest to maintain state of the client's workspace and instead create a copy and delete it when done.

Reply:
I think I was imagining it could also be used to augment the `githubrepo` client. Although that downloads via the tarball, so the `.git` folder wouldn't be there.

Reply:
Resolved after discussion of which role the git client plays. Not a top-level RepoClient implementation, but rather to potentially replace `tarballHandler`